File: app/src/main/java/com/kelsos/mbrc/services/NowPlayingServiceImpl.java
package com.kelsos.mbrc.services;

import android.support.annotation.NonNull;

import com.fasterxml.jackson.core.type.TypeReference;
import com.kelsos.mbrc.constants.Protocol;
import com.kelsos.mbrc.data.NowPlaying;
import com.kelsos.mbrc.data.Page;
import com.kelsos.mbrc.data.PageRange;
import com.kelsos.mbrc.data.SocketMessage;

import java.io.IOException;

import rx.Observable;

public class NowPlayingServiceImpl extends ServiceBase implements NowPlayingService {
  @Override
  public Observable<Page<NowPlaying>> getNowPlaying(int offset, int limit) {
    PageRange range = getPageRange(offset, limit);
    return request(Protocol.NowPlayingList,  range == null ? "" : range).flatMap(this::getPageObservable);
  }

  @NonNull
  private Observable<Page<NowPlaying>> getPageObservable(SocketMessage socketMessage) {
    return Observable.create(subscriber -> {
      try {
        TypeReference<Page<NowPlaying>> typeReference = new TypeReference<Page<NowPlaying>>() {
        };
        Page<NowPlaying> page = mapper.readValue((String) socketMessage.getData(), typeReference);
        subscriber.onNext(page);
        subscriber.onCompleted();
      } catch (IOException e) {
        subscriber.onError(e);
      }
    });
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/services/NowPlayingServiceImpl.kt
package com.kelsos.mbrc.services

import com.fasterxml.jackson.databind.ObjectMapper
import com.kelsos.mbrc.constants.Protocol
import com.kelsos.mbrc.data.NowPlaying
import com.kelsos.mbrc.data.Page
import com.kelsos.mbrc.extensions.toPage
import com.kelsos.mbrc.repository.ConnectionRepository
import com.kelsos.mbrc.utilities.SettingsManager
import io.reactivex.Observable
import javax.inject.Inject

class NowPlayingServiceImpl
@Inject
constructor(
    repository: ConnectionRepository,
    private val mapper: ObjectMapper,
    settingsManager: SettingsManager
) : ServiceBase(repository, mapper, settingsManager), NowPlayingService {

  override fun getNowPlaying(offset: Int, limit: Int): Observable<Page<NowPlaying>> {
    val range = getPageRange(offset, limit)
    return request(Protocol.NowPlayingList, range).flatMap { it.toPage<NowPlaying>(mapper) }
  }
}
